### PR TITLE
WebGPURenderer: Build - `three.webgpu.js`

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/pull/28328, https://github.com/mrdoob/three.js/pull/28328#issuecomment-2136541224

**Description**

### TSL

I did some tests to include TSL in the build, the first attempt was `THREE.TSL` and use `const { .. } = ` to import, which works well but is incompatible with `tree shaking`, the second attempt was to preserve `three/tls` but that would prevent having a single build file for everything. I think that since TSL will be part of the agnostic rendering system, it seems correct for it to be an essential central part of the file, just like other materials are.

To import a TSL, just use:

```
import { float, vec3, color } from 'three/tsl';
```

For class creation this is more familiar:

```js
const material = THREE.MeshStandardNodeMaterial();
```

### Renderer

Using the Renderer has become simpler too with `THREE.`:

```js
const renderer = new THREE.WebGPURenderer();
```
> One possibility is that we can keep a `WebGLRenderer` class using `forceWebGL` automatically.

### PMREMGenerator

Previously it was necessary to reimport PMREMGenerator from a new path for webgpu, now it is possible to use the original `THREE.PMREMGenerator` syntax.

```js
const sceneRT = new THREE.PMREMGenerator( renderer ).fromScene( scene );
```

### RectAreaLightTexturesLib

I created a base LTC for use on WebGPU, WebGLRenderer also relies on this same class.

### QuadMesh, PostProcessing

QuadMesh and PostProcessing can be used with THREE.QuadMesh and THREE.PostProcessing.

### Lines2

We have a separate `jsm/lines/webgpu` class now to handle just TSL, previously we had a lot of GLSL mixed in between the class extensions.

### Imports

```json
{
	"imports": {
		"three": "../build/three.webgpu.js",
		"three/debug": "../src/Three.WebGPU.js",
		"three/addons/": "./jsm/"
	}
}
```

If you want to edit the files without having to recompile them, you can use the imports above.

## Next steps

This simplifies the process for creating the `tree shaking` that will come with the next PR. Do you just need to update from `three.webgpu.js` to `three.webgpu.nodes.js` or `three.nodes.js`?

### CODE REVIEW

**Comment by sunag** [comment]
Seems that an update in `puppeteer` will be necessary.

**Comment by RenaudRohlinger** [comment]
Could we also use this PR to move the `WebGPURenderer` and `TSL` files to a directory like `/webgpu/`? 😄

It feels a bit odd to point to the addons such as `examples/jsm/renderers/webgpu/Three.js` for the build. This change would also make it easier for maintainers to work on new the Renderer by separate the TSL and WebGPURenderer files from the hundreds of existing WebGLRenderer examples blended in.

The import map would then look like this:
```js
{
	"imports": {
		"three": "../build/three.webgpu.js",
		"three/debug": "./webgpu/Three.js",
		"three/addons/": "./jsm/"
	}
}
```

/cc @mrdoob @sunag @Mugen87

**Comment by Mugen87** [comment]
It wouldn't be `examples/jsm/renderers/webgpu/Three.js` but `three/addons/renderers/webgpu/Three.js` when importing from the npm package. 

TBH, I did not have the impression so far that import paths containing `addons` are an issue. Moving around source files in directories is also a heavy breaking change and I would like to avoid it whenever possible unless there is a good reason (which I don't see right now).

**Comment by RenaudRohlinger** [comment]
> TBH, I did not have the impression so far that import paths containing addons are an issue.

It's a matter of best practice and trying to organize the repository more effectively. Why should the new Renderer and Node System be under the "examples/jsm" folder?
I understand that historically it was done this way, and while I don't mind maintaining that approach, it does seem counterintuitive.

It feels like anything that is not the WebGLRenderer is placed under examples/jsm by default.

**Comment by Mugen87** [comment]
> It feels like anything that is not the WebGLRenderer is placed under examples/jsm by default.

Um, why not moving `WebGPURenderer` and TSL related modules into `src`? Wouldn't that be the more obvious approach?

The directory size of `src` or its number of files is not the critical issue. It's the size of the build files that should as small as possible. It should be possible to manage more code in `src` while keeping the builds sizes consistent.

**Comment by RenaudRohlinger** [comment]
Indeed, even better! 👍

File: examples/webgpu_sandbox.html
Comment (WestLangley) [inline examples/webgpu_sandbox.html:18]: `debug` does not appear to be needed. Should we have an example in which it is?

Comment (sunag) [reply examples/webgpu_sandbox.html:18]: I purposely put it in `sandbox` because I thought it could convey the idea of ​​`debug` since I can't make simple comments in the JSON, maybe there is a more suitable way to do this...

Comment (WestLangley) [reply examples/webgpu_sandbox.html:18]: Your choice... I'd vote to merge this PR ASAP if @Mugen87 approves.

Comment (mrdoob) [reply examples/webgpu_sandbox.html:18]: Let me think about the naming for all this for a couple of days 🙏

Comment (sunag) [reply examples/webgpu_sandbox.html:18]: I replaced `three/nodes` to `three/webgpu`. I imagined that developers could use three alias in vite like:

```js
// vite.config.js

import { defineConfig } from 'vite';

export default defineConfig({
  resolve: {
    alias: {
      'three': 'three/webgpu'
    }
  }
});
```

**Comment by mrdoob** [comment]
> I did some tests to include TSL in the build, the first attempt was `THREE.TSL` and use `const { .. } = ` to import, which works well but is incompatible with `tree shaking`

Hmm, that would definitely more clean than `import { float, vec3, color } from 'three';`...

Why is it incompatible with `tree shaking`?

**Comment by sunag** [comment]
> Why is it incompatible with tree shaking?

I created a mini example in the link below, when I used `const { .. } = `, it imports all the values ​​from TSL, unlike when I use `THREE.TSL.vec3()` for example.

https://shorturl.at/FWZOC

**Comment by sunag** [comment]
> I created a mini example in the link below, when I used const { .. } = , it imports all the values ​​from TSL, unlike when I use THREE.TSL.vec3() for example.

Same thing with `vite`.

Maybe we could add a `three/tsl` alias and advance? I'm a little worried about revisiting this PR every time we merge some PR, it feels like it's going to break at some point.

**Comment by mrdoob** [comment]
> Maybe we could add a `three/tsl` alias and advance?

Yes, let's try that first 👍

**Review by Mugen87 — APPROVED** [review]

**Comment by RenaudRohlinger** [comment]
For documentation purpose here is the correct way to use the new build:
```js
// vite.config.js

import { defineConfig } from 'vite';

export default defineConfig({
  resolve: {
    alias: {
       'three/addons': 'three/examples/jsm',
       'three/tsl': 'three/webgpu',
       'three': 'three/webgpu',
    }
  }
});
```

The order is important otherwise `three/tsl` will try to resolve `three/webgpu/tsl` for example.

/cc @sunag

**Pull request merged**
